src/gurt/telemetry.c

```diff
@@ -1981,173 +1981,76 @@ d_tm_get_version(void)
 }
 
 /**
- * Perform a directory listing at the path provided for the items described by
- * the \a d_tm_type bitmask.  A result is added to the list it if matches
- * one of the metric types specified by that filter mask. The mask may
- * be a combination of d_tm_metric_types.  The search is performed only on the
- * direct children specified by the path.  Returns a linked list that points to
+ * Perform a recursive directory listing from the given \a node for the items
+ * described by the \a d_tm_type bitmask.  A result is added to the list if it
+ * matches one of the metric types specified by that filter mask. The mask may
+ * be a combination of d_tm_metric_types.  Creates a linked list that points to
  * each node found that matches the search criteria.  Adds elements to an
  * existing node list if head is already initialized. The client should free the
  * memory with d_tm_list_free().
  *
  * \param[in,out]	head		Pointer to a nodelist
  * \param[in]		shmem_root	Pointer to the shared memory segment
- * \param[in]		path		Full path name to the directory or
- *					metric to list
+ * \param[in]		node		The recursive directory listing starts
+ *					from this node.
  * \param[in]		d_tm_type	A bitmask of d_tm_metric_types that
  *					filters the results.
  *
  * \return		D_TM_SUCCESS		Success
  *			-DER_NOMEM		Out of global heap
- *			-DER_EXCEEDS_PATH_LEN	The full name length is
- *						too long
- *			-DER_METRIC_NOT_FOUND	No items were found at the path
- *						provided
  *			-DER_INVAL		Invalid pointer for \a head or
- *						\a path
+ *						\a node
  */
 int
-d_tm_list(struct d_tm_nodeList_t **head, uint64_t *shmem_root, char *path,
-	  int d_tm_type)
+d_tm_list(struct d_tm_nodeList_t **head, uint64_t *shmem_root,
+	  struct d_tm_node_t *node, int d_tm_type)
 {
-	struct d_tm_nodeList_t	*nodelist = NULL;
-	struct d_tm_node_t	*node = NULL;
-	struct d_tm_node_t	*parent_node = NULL;
-	char			*str = NULL;
-	char			*token;
-	char			*rest;
-	int			rc = D_TM_SUCCESS;
-	bool			search_siblings = false;
+	int	rc = D_TM_SUCCESS;
 
-	if ((head == NULL) || (path == NULL)) {
+	if ((head == NULL) || (node == NULL)) {
 		rc = -DER_INVAL;
 		goto failure;
 	}
 
-	D_STRNDUP(str, path, D_TM_MAX_NAME_LEN);
-
-	if (str == NULL) {
-		rc = -DER_NOMEM;
-		D_ERROR("Failed to allocate memory for path [%s]: rc = %d\n",
-			path, rc);
-		goto failure;
+	if (d_tm_type & node->dtn_type) {
+		rc = d_tm_add_node(node, head);
+		if (rc != D_TM_SUCCESS)
+			goto failure;
```

Comment:
Would it be interesting to log these failure cases? Or can this stuff happen under normal circumstances, such that it would be noisy?

Reply:
The failure should not occur often.  There are two reasons why it might fail, 1) system ran out of memory, 2) developer passed in bad arguments.  I don't think that logging it would be noisy, but I wrote it so that it returns the error code so that a developer could read that while developing the code and take necessary action.  If there is consensus on a better approach, I'll retool that as necessary.